tests/ModelContextProtocol.Tests/Client/McpClientExtensionsTests.cs

```diff
@@ -376,4 +380,125 @@ public async Task WithDescription_ChangesToolDescription()
+
+        // Assert
+        Assert.NotNull(notification.Params);
+        // Parse the Params string back to a CancelledNotification
+        var cancelled = JsonSerializer.Deserialize<CancelledNotification>(notification.Params.ToString());
+        Assert.NotNull(cancelled);
+        Assert.Equal(rpcNotification.RequestId.ToString(), cancelled.RequestId.ToString());
+        Assert.Equal(rpcNotification.Reason, cancelled.Reason);
+    }
+
+    [Fact]
+    public async Task Should_Not_Intercept_Sent_Notifications()
+    {
+        // Arrange
+        var token = TestContext.Current.CancellationToken;
+        TaskCompletionSource<JsonRpcNotification> clientReceived = new();
+        await using var client = await CreateMcpClientForServer(
+            options: CreateClientOptions([new(NotificationMethods.CancelledNotification, notification =>
+            {
+                var exception = new InvalidOperationException("Should not intercept sent notifications");
+                clientReceived.TrySetException(exception);
+                return clientReceived.Task;
+            })]),
+            cancellationToken: token);
+
+        // Act
+        await client.NotifyCancelAsync(
+            requestId: new("abc"),
+            reason: "Cancelled",
+            cancellationToken: token);
+        await Assert.ThrowsAsync<TimeoutException>(
+            async () => await clientReceived.Task
+                .WaitAsync(TimeSpan.FromSeconds(5), token));
+        // Assert
+        Assert.False(clientReceived.Task.IsCompleted);
+    }
+
+    [Fact]
+    public async Task Can_Notify_Cancel()
+    {
+        // Arrange
+        var token = TestContext.Current.CancellationToken;
+        await using var client = await CreateMcpClientForServer(
+            cancellationToken: token);
+        RequestId expectedRequestId = new("abc");
+        var expectedReason = "Cancelled";
+
+        // Act
+        await client.NotifyCancelAsync(
+            requestId: expectedRequestId,
+            reason: expectedReason,
+            cancellationToken: token);
+
+        // Assert
+        var result = await _clientToServerPipe.Reader.ReadAsync(token);
+        var copyBytes = new byte[result.Buffer.Length];
+        result.Buffer.CopyTo(copyBytes);
+        Utf8JsonReader reader = new(copyBytes);
+        var jsonRpcNotification = JsonSerializer.Deserialize<JsonRpcNotification>(ref reader);
+        Assert.NotNull(jsonRpcNotification);
+        Assert.Equal(NotificationMethods.CancelledNotification, jsonRpcNotification.Method);
+
+        var parameters = jsonRpcNotification.Params;
+        Assert.NotNull(parameters);
+        var cancelledNotification = JsonSerializer.Deserialize<CancelledNotification>(parameters.ToString());
+        Assert.NotNull(cancelledNotification);
+        Assert.Equal(expectedRequestId.ToString(), cancelledNotification.RequestId.ToString());
+        Assert.Equal(expectedReason, cancelledNotification.Reason);
```

Comment:
This shouldn't need to be interacting with _clientToServerPipe manually, manually deserializing JsonRpcNotifications, etc. It should be able to do something like the tests in https://github.com/modelcontextprotocol/csharp-sdk/pull/223, just creating the server with a notification handler that sets a TaskCompletionSource when the cancellation notification arrives, sending such a notification on the client, and awaiting that TCS.Task.